### DISCUSSION TITLE
Moving functions / structs / parameters

### DISCUSSION THREAD

**Discussion by leandrobbraga** [helix-editor/helix#8467] [Ideas]
Hello, what do you guys think about the possibility of moving functions, structs and parameters. Is it currently possible?

Imagine that there are three functions
```text
f_1
f_2
f_3 <- cursor here
```

Would be useful to have the possibility of swapping their positions around (move-up, move-down).

For example, after a `move-function-up`:
```text
f_1
f_3 <- cursor here
f_2
```

The same would go for structs and function parameters, functions and struct being interchangeable with each other.

I used to use this feature all the time in Intellij.

I guess this would be possible, since helix is based on tree sitter.

Thank you!

**Comment by iceghost** [comment]
For lines, I use `x d kkkjjjkjkj P`.

Occasionally, for functions, I do `maf xd kkkkkjjkkjjkkk P`.

**Reply by leandrobbraga** [reply]
Is it possible to bind "maf"? I couldn't find it.

I manage to build a keybind that does `["delete_selection", "goto_next_function", "add_newline_below", "move_line_down", "paste_after", "add_newline_below"]` and another that does `["delete_selection", "goto_prev_function", "move_line_up", "paste_before", "add_newline_below"]`. This is very close, but not quite, just missing the "maf" part.

**Reply by gabydd** [reply]
Not currently (https://github.com/helix-editor/helix/pull/1723 and https://github.com/helix-editor/helix/pull/4287)

**Reply by leandrobbraga** [reply]
Thank you, I guess I'll have everything I need after those are merged. So I'm not 100% sure if it's necessary to have this function out-of-the-box since it will be possible to build it manually

**Reply by leandrobbraga** [reply]
This solution doesn't quite work in all situations, it does not correct for indentation for example.

**Reply by kirawi** [reply]
See https://github.com/helix-editor/helix/issues/6044
